### DISCUSSION TITLE
`AsyncSequence` Dependencies

### DISCUSSION THREAD

**Discussion by rhysm94** [pointfreeco/swift-composable-architecture#1568] [Q&A]
Hey all,

Perhaps slightly off topic, but I think it's on-topic to how dependencies are typically defined when using TCA.

Just wondering if anyone has a dependency which returns an `AsyncSequence`, and if so, how you're handling the fact it's missing a primary associated type?

For example, I'm using GRDB as my database layer in this application. There's a certain table, say of `Player` types, that I want to observe. So in my `DatabaseManager` type, I would implement my `observePlayers` method like this:

```swift
ValueObservation.tracking(Player.fetchAll).values(in: writer)
```

However, this returns a GRDB specific type - `AsyncValueObservation<Player>`. Obviously, I don't want any GRDB types to bleed into my `DatabaseManager` interface, as I don't want to be building GRDB each time I build my non-live `DatabaseManager` implementation (even though it's a small dependency and doesn't take long to build).

If I was doing this with Combine (which I probably will go back to), the type for my `observePlayers` method would be `@Sendable () -> AnyPublisher<[Player], Error>`. Ideally, I'd like to do quite similar with `async`/`await` instead of Combine.

Is there a way around this? How are you designing `AsyncSequence`-returning dependencies?

**Comment by mbrandonw** [comment]
You can _kinda_ think of `AsyncStream` as a type erased `AsyncSequence` (with caveats), and in fact the library ships with an initializer on `AsyncStream` that takes a sequence to do the erasing:

https://github.com/pointfreeco/swift-composable-architecture/blob/ed3a380f09a81e72636d7b0699bf2dd2e6313780/Sources/ComposableArchitecture/Effects/ConcurrencySupport.swift#L1-L76

We have an example of using this in the long-living effect case study, where we erase the `NotificationCenter` async sequence to be an `AsyncStream`:

https://github.com/pointfreeco/swift-composable-architecture/blob/ed3a380f09a81e72636d7b0699bf2dd2e6313780/Examples/CaseStudies/SwiftUICaseStudies/02-Effects-LongLiving.swift#L57-L63

The biggest caveat is that `AsyncStream` does not support multiple subscribers. So, it only works if you are willing to vend a fresh `AsyncStream` for each time your dependency endpoint is called. That's why the screenshot dependency above is a function `() -> AsyncStream` rather than a bare `AsyncStream`.

**Reply by tgrapperon** [reply]
There is another caveat with the `AsyncStream` when initialized from any `AsyncSequence`, as it starts observing the sequence as soon as it is created and will replay these values when you start iterating over it. So it is better to create this value at the last moment, and not store it around to start iterating later.

**Reply by JacksonUtsch** [reply]
These caveats make it hard for usage. Especially no `CurrentValueSubject` equivalent without using a third party `AsyncSequence` implementation. Hoping good native support comes soon, meanwhile I'll be sticking to `Combine`.

**Reply by rhysm94** [reply]
> These caveats make it hard for usage. Especially no `CurrentValueSubject` equivalent without using a third party `AsyncSequence` implementation. Hoping good native support comes soon, meanwhile I'll be sticking to `Combine`.

I believe there's an `AsyncChannel` type in Async Algorithms from Apple, but iirc, it's more like a `PassthroughSubject` than a `CurrentValueSubject`. Is that right?

**Reply by tgrapperon** [reply]
`AsyncChannel` has the same "Unique consumer" issue than any `AsyncSequence` unfortunately.
The potential replacement will likely involve [`AsyncBroadcastSequence`](https://github.com/apple/swift-async-algorithms/pull/214) to allow multiple consumers like `CurrentValueSubject`, but I'm afraid we'll have to wait for a long time for official support.

**Reply by corysullivan** [reply]
You can also have a look at [AsyncExtensions](https://github.com/sideeffect-io/AsyncExtensions). There is some overlap with apples Async Algorithms, but it seems to have some useful operators like `multicast`  and a  `AsyncCurrentValueSubject`
